### ISSUE TITLE
[Feature Request]: type safe ipc handling

### ISSUE THREAD

**Issue by Juice10** [electron/electron#33691]
### Preflight Checklist

- [X] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/main/CONTRIBUTING.md) for this project.
- [X] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/electron/electron/issues) for a feature request that matches the one I want to file, without success.

### Problem Description

In typescript when using methods like `ipcMain.handle`, `ipcRenderer.invoke` and [other](https://www.electronjs.org/docs/latest/api/ipc-main/#methods) [methods],(https://www.electronjs.org/docs/latest/api/ipc-renderer#methods) its hard to make sure they are type safe and the request and response interfaces are kept in sync.

[There are a number of external packages trying to solve this issue](https://www.npmjs.com/search?ranking=optimal&q=typescript%20electron%20ipc) but after having tried a couple, they all seem either too overbearing, unmaintained, incomplete or a combination of the above.

### Proposed Solution

It would be great if IPC methods would accept a generic type which allow for better type safety. 
(Inspired by the incomplete [electron-typescript-ipc](https://www.npmjs.com/package/electron-typescript-ipc) npm package)

```ts
ipcMain.handle<T>(...)
``` 

### Alternatives Considered

Alternatively Electron could "pick a winner" from all the [potential candidates](https://www.npmjs.com/search?ranking=optimal&q=typescript%20electron%20ipc), and advertise it in the docs. That would focus and hopefully consolidate the landscape, and make it easier for the electron/typescript community to maintain one project instead of 25.

### Additional Information

https://stackoverflow.com/questions/61767919/how-can-i-make-electron-channels-more-type-safe

**Comment by MarshallOfSound** [comment]
> ipcMain.handle<T>(...)

This is fake type safety, there is no runtime guarantee that the IPC message received has that structure and when the structure of the incoming message matters not only for stability purposes but also for security purposes.

I know this is probably going to sound like "yet another implementation" but I would recommend using using a combination of [`joi`](https://www.npmjs.com/package/joi) and [`joiful`](https://github.com/joiful-ts/joiful) to obtain static typing and runtime type safety.

I've looked through all the potential candidates you linked and I've seen most of them before, which is what has led me to write my own solution multiple times 😆 I'll leave this issue open because I have some ideas in this space, but nothing near-term for sure.

**Comment by Juice10** [comment]
> > This is fake type safety
> 
> 
> 
> My impression is that @Juice10 was referring to the typescript/dev-time type-safety, at least based on its multiple mentions.

Yes that correct, I'm more concerned with keeping bugs to a minimum than minimizing attack vector (in this case).

> > Alternatively Electron could "pick a winner" from all the potential candidates, and advertise it in the docs.
> 
> 
> 
> I think advertising in the docs won't be enough as this won't be a very forward-looking nor sustainable scenario but accepting the burden of maintaining the library in @electron org.
If an external tool or library ends up being a solution than getting it fully adopted in the electron project would be ideal indeed!

**Comment by mckravchyk** [comment]
I look at the "potential candidates" and most if not all of those packages are not actually typed IPC. Let me self-promo my own [electron-ipc-extended)[https://github.com/mckravchyk/electron-ipc-extended]. Besides typed IPC, it also enables invoking commands from main process to renderers.

**Comment by tanhh326** [comment]
```ts
// @types/electron/index.d.ts
import "electron";

declare global {
  namespace Electron {
    interface IpcMain {
      on(channel: "abs", listener: (event: IpcMainEvent, arg: string, arg1: string) => void): this;
    }
  }
}
```

Is custom type file overwriting a feasible solution?

**Comment by Wiktor102** [comment]
Could we at least get type info for `ipcRenderer`? Currently, it's `any`, so even no method names autocomplete.
![Image](https://github.com/user-attachments/assets/d4932403-74d4-4ce3-b363-a870bdbeb653)

__Some context:__
Just now, I've run into a bug because I used `send` instead of `invoke`, and due to missing types, TypeScript didn't show an error (I was returning the result of `send` where my method was expecting `Promise<ISomething>`).

__PS:__ I'm new to Electron, hope I am not just missing an import?
